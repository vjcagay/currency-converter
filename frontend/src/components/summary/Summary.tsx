import React from "react";
import styles from "./styles.module.css";

type Props = {
  className?: string;
  from: string;
  to: string;
  firstRateValue: number;
  lastRateValue: number;
  rangeString?: string;
};

const currencyTickerNames: { [key: string]: string } = {
  USD: "US Dollar",
  SGD: "Singapore Dollar",
  JPY: "Japanese Yen",
  EUR: "Euro",
};

const GainLossPercentage = (props: { value1: number; value2: number }) => {
  const gainLoss = ((props.value2 - props.value1) / props.value2) * 100;
  const isPositive = gainLoss > 0;

  const result = `${isPositive ? "+" : ""}${gainLoss.toFixed(3)}%`;

  return <span className={`${styles.gainLossText} ${isPositive ? styles.gain : styles.loss}`}>{result}</span>;
};

const Summary = (props: Props) => {
  return (
    <div className={`${styles.wrapper} ${props.className || ""}`}>
      <b className={styles.titleText}>
        {props.from} to {props.to} Chart
      </b>
      <wbr />
      &nbsp;
      <GainLossPercentage value1={props.firstRateValue} value2={props.lastRateValue} />
      &nbsp;
      {props.rangeString && <span className={styles.rangeText}>({props.rangeString})</span>}
      <br />
      <small>
        {currencyTickerNames[props.from]} to {currencyTickerNames[props.to]}
      </small>
    </div>
  );
};

export default Summary;
