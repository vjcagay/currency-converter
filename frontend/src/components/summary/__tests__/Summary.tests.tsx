import { render } from "@testing-library/react";
import React from "react";
import Summary from "../Summary";

describe("<Summary />", () => {
  it("should render correctly with a gain percentage", () => {
    const { asFragment } = render(
      <Summary from="USD" to="JPY" firstRateValue={146.942966} lastRateValue={147.470418} rangeString="1W" />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render correctly with a loss percentage", () => {
    const { asFragment } = render(
      <Summary from="USD" to="JPY" firstRateValue={147.470418} lastRateValue={146.942966} rangeString="1W" />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
