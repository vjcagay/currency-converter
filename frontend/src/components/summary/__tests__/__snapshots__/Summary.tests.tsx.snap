// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Summary /> should render correctly with a gain percentage 1`] = `
<DocumentFragment>
  <div
    class="wrapper "
  >
    <b
      class="titleText"
    >
      USD to JPY Chart
    </b>
    <wbr />
     
    <span
      class="gainLossText gain"
    >
      +0.358%
    </span>
     
    <span
      class="rangeText"
    >
      (1W)
    </span>
    <br />
    <small>
      US Dollar to Japanese Yen
    </small>
  </div>
</DocumentFragment>
`;

exports[`<Summary /> should render correctly with a loss percentage 1`] = `
<DocumentFragment>
  <div
    class="wrapper "
  >
    <b
      class="titleText"
    >
      USD to JPY Chart
    </b>
    <wbr />
     
    <span
      class="gainLossText loss"
    >
      -0.359%
    </span>
     
    <span
      class="rangeText"
    >
      (1W)
    </span>
    <br />
    <small>
      US Dollar to Japanese Yen
    </small>
  </div>
</DocumentFragment>
`;
